src/uu/touch/src/touch.rs

```diff
@@ -165,8 +165,11 @@ fn is_first_filename_timestamp(
         // env check is last as the slowest op
         && matches!(std::env::var("_POSIX2_VERSION").as_deref(), Ok("199209"))
     {
-        let s = files[0];
-        all_digits(s) && (s.len() == 8 || (s.len() == 10 && (69..=99).contains(&get_year(s))))
+        if let Some(s) = files[0].to_str() {
+            all_digits(s) && (s.len() == 8 || (s.len() == 10 && (69..=99).contains(&get_year(s))))
+        } else {
+            false
+        }
     } else {
         false
     }
```

Comment:
This looks a bit weird ;-)

A functional approach might be cleaner (with removing the `if`), something like
```rust
    && files[0].to_str().map_or(false, |s| all_digits(s) && (s.len() == 8 || (s.len() == 10 && (69..=99).contains(&get_year(s)))))
}
```
Or with moving the closure into its own function:
```rust
    && files[0].to_str().map_or(false, is_timestamp)
}

fn is_timestamp(s: &str) -> bool {
    all_digits(s) && (s.len() == 8 || (s.len() == 10 && (69..=99).contains(&get_year(s))))
}
```